### PR TITLE
Implement bits of BaseHeightMapRenderObjClass, FlatHeightMapRenderObjClass and HeightMapRenderObjClass

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/flatheightmap.cpp [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+void FlatHeightMapRenderObjClass::Adjust_Terrain_LOD(int adj)
+{
+    BaseHeightMapRenderObjClass::Adjust_Terrain_LOD(adj);
+}
+
+void FlatHeightMapRenderObjClass::Do_Partial_Update(
+    IRegion2D &partial_range, WorldHeightMap *map, RefMultiListIterator<RenderObjClass> *lights)
+{
+    if (map != nullptr) {
+        Ref_Ptr_Release(m_map);
```
Comment (tomsons26) [inline src/game/client/flatheightmap.cpp]: Ref_Ptr_Set i believe

File: src/game/client/baseheightmap.cpp [hunk 2]
```diff
@@ -71,32 +506,436 @@ float BaseHeightMapRenderObjClass::Get_Height_Map_Height(float x, float y, Coord
...
+}
+
+bool BaseHeightMapRenderObjClass::Evaluate_As_Visible_Cliff(int x, int y, float height)
+{
+    static float distance[4] = { 0.0f, 10.0f, GameMath::Sqrt(2.0f) * 10.0f, 10.0f };
+    unsigned char heights[4];
+    float fheights[4];
+
+    heights[0] = m_map->Get_Height(x, y);
+    heights[0] = m_map->Get_Height(x + 1, y);
```
Comment (tomsons26) [inline src/game/client/baseheightmap.cpp]: all aren't 0 but sequential

File: src/game/client/baseheightmap.cpp [hunk 3]
```diff
@@ -13,12 +13,446 @@
...
+void BaseHeightMapRenderObjClass::Update_Center(CameraClass *camera, RefMultiListIterator<RenderObjClass> *lights)
+{
+    if (m_map != nullptr && !m_updating) {
+        if (m_treeBuffer != nullptr) {
+            m_treeBuffer->Do_Full_Update();
+        }
+
+        if (m_propBuffer != nullptr) {
+            m_propBuffer->Update_Center();
+        }
```
Comment (tomsons26) [inline src/game/client/baseheightmap.cpp:291]: missing code

File: src/game/client/flatheightmap.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/flatheightmap.cpp]: fixed.

File: src/game/client/baseheightmap.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/baseheightmap.cpp]: fixed.

File: src/game/client/baseheightmap.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/baseheightmap.cpp:291]: fixed.

File: src/game/client/baseheightmap.cpp [hunk 4]
```diff
@@ -71,32 +519,436 @@ float BaseHeightMapRenderObjClass::Get_Height_Map_Height(float x, float y, Coord
...
+    if (m_propBuffer != nullptr) {
+        m_propBuffer->Add_Prop(id, location, orientation, scale, name);
+    }
+}
+
+void BaseHeightMapRenderObjClass::Add_Scorch(Vector3 location, float radius, Scorches type)
+{
+    if (m_numScorches >= 500) {
+        for (int i = 0; i < 499; i++) {
+            this->m_scorches[i] = this->m_scorches[i + 1];
```
Comment (tomsons26) [inline src/game/client/baseheightmap.cpp]: this

File: src/game/client/baseheightmap.cpp [hunk 5]
```diff
@@ -71,32 +519,436 @@ float BaseHeightMapRenderObjClass::Get_Height_Map_Height(float x, float y, Coord
...
+        }
```
Comment (tomsons26) [inline src/game/client/baseheightmap.cpp:554]: missing --m_numScorches

File: src/game/client/baseheightmap.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/baseheightmap.cpp]: fixed.

File: src/game/client/baseheightmap.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/baseheightmap.cpp:554]: fixed.

File: src/game/client/baseheightmap.cpp [hunk 6]
```diff
@@ -13,12 +13,459 @@
...
+        g_theSnowManager->Re_Acquire_Resources();
+    }
+
+    if (m_roadBuffer != nullptr) {
+        m_roadBuffer->Allocate_Road_Buffers();
+        m_roadBuffer->Load_Roads();
+    }
+    if (g_theTacticalView != nullptr) {
+        g_theTacticalView->Force_Redraw();
+        g_theTacticalView->Force_Redraw();
```
Comment (tomsons26) [inline src/game/client/baseheightmap.cpp]: comment as a possible bug

Comment (jonwil) [reply src/game/client/baseheightmap.cpp]: fixed.

File: src/game/client/flatheightmap.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        region.lo.y = 0;
+        region.hi.x = m_map->Get_X_Extent();
+        region.hi.y = m_map->Get_Y_Extent();
+
+        for (int i = 0; i < m_tileWidth; i++) {
+            for (int j = 0; j < m_tileHeight; j++) {
+                m_tiles[m_tileWidth * j + i].Do_Partial_Update(region, m_map, true);
+            }
+        }
+    }
```
Comment (tomsons26) [inline src/game/client/flatheightmap.cpp:224]: should call base reaquire

Comment (jonwil) [reply src/game/client/flatheightmap.cpp:224]: fixed.

File: src/game/client/heightmap.cpp [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        m_vertexBufferTiles = 0;
+    }
+
+    if (m_vertexBufferBackup != nullptr) {
+        for (int i = 0; i < m_numVertexBufferTiles; i++) {
+            delete[] m_vertexBufferBackup[i];
+        }
+
+        delete[] m_vertexBufferBackup;
+        m_vertexBufferBackup = 0;
```
Comment (xezon) [inline src/game/client/heightmap.cpp]: nullptr

File: src/game/client/flatheightmap.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    int m_tileWidth;
+    int m_tileHeight;
+    int m_state;
+};
+
+#ifdef GAME_DLL
+extern FlatHeightMapRenderObjClass *&g_theFlatHeightMap;
+#else
+extern FlatHeightMapRenderObjClass *g_theFlatHeightMap;
+#endif
```
Comment (xezon) [inline src/game/client/flatheightmap.h]: eol

File: src/game/client/heightmap.h [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        int origin_x,
+        int origin_y,
+        W3DDynamicLight **lights,
+        int num_lights);
+    void Render_Terrain_Pass(CameraClass *camera);
+    void Render_Extra_Blend_Tiles();
+
+private:
+    unsigned int *m_extraBlendTilePositions;
+    int m_extraBlendTileCount;
```
Comment (tomsons26) [inline src/game/client/heightmap.h]: m_numExtraBlendTiles

File: src/game/client/heightmap.h [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    int m_extraBlendTilesRendered;
+    int m_extraBlendTileSize;
```
Comment (tomsons26) [inline src/game/client/heightmap.h]: m_extraBlendTilePositionsSize

File: src/game/client/heightmap.h (hunk 10 above)
Comment (jonwil) [reply src/game/client/heightmap.h]: fixed.

File: src/game/client/heightmap.h (hunk 11 above)
Comment (jonwil) [reply src/game/client/heightmap.h]: fixed.

File: src/game/client/baseheightmap.cpp [hunk 12]
```diff
@@ -13,12 +13,459 @@
...
+        g_theSnowManager->Re_Acquire_Resources();
+    }
+
+    if (m_roadBuffer != nullptr) {
+        m_roadBuffer->Allocate_Road_Buffers();
+        m_roadBuffer->Load_Roads();
+    }
+    if (g_theTacticalView != nullptr) {
+        g_theTacticalView->Force_Redraw();
+        g_theTacticalView->Force_Redraw(); // possible bug
```
Comment (xezon) [inline src/game/client/baseheightmap.cpp]: Add a "TODO" keyword to make it discoverable.

File: src/game/client/heightmap.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/heightmap.cpp]: fixed.

File: src/game/client/flatheightmap.h (hunk 9 above)
Comment (jonwil) [reply src/game/client/flatheightmap.h]: fixed.

File: src/game/client/baseheightmap.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/baseheightmap.cpp]: fixed.

File: src/game/client/baseheightmap.h [hunk 13]
```diff
@@ -111,6 +194,30 @@ class BaseHeightMapRenderObjClass : public RenderObjClass, public DX8_CleanupHoo
         }
     }
 
+    unsigned char Get_Clip_Height(int x, int y)
+    {
+        int x_extent = m_map->Get_X_Extent() - 1;
+        int y_extent = m_map->Get_Y_Extent() - 1;
+
+        if (x >= 0) {
```
Comment (tomsons26) [inline src/game/client/baseheightmap.h]: flip

File: src/game/client/baseheightmap.h [hunk 14]
```diff
@@ -111,6 +194,30 @@ class BaseHeightMapRenderObjClass : public RenderObjClass, public DX8_CleanupHoo
...
+            if (x > x_extent) {
+                x = x_extent;
+            }
+        } else {
+            x = 0;
+        }
+
+        if (y >= 0) {
```
Comment (tomsons26) [inline src/game/client/baseheightmap.h]: flip

File: src/game/client/baseheightmap.h (hunk 13 above)
Comment (jonwil) [reply src/game/client/baseheightmap.h]: fixed.

File: src/game/client/baseheightmap.h (hunk 14 above)
Comment (jonwil) [reply src/game/client/baseheightmap.h]: fixed.

File: src/game/client/heightmap.cpp [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+void HeightMapRenderObjClass::Render_Terrain_Pass(CameraClass *camera)
+{
+#ifdef BUILD_WITH_D3D8
+    Matrix3D tm(true);
+    DX8Wrapper::Set_Transform(D3DTS_WORLD, tm);
+    DX8Wrapper::Set_Index_Buffer(m_indexBuffer, 0);
+
+    for (int i = 0; i < m_numVBTilesY; i++) {
+        for (int j = 0; j < m_numVBTilesX; j++) {
+            static int count = 0;
```
Comment (tomsons26) [inline src/game/client/heightmap.cpp]: needs to be at top else it sets it every loop

Comment (jonwil) [reply src/game/client/heightmap.cpp]: fixed.

**Comment by xezon** [comment]
Heightmaps appear to look as usual. No crashes. Tested in RelWithDebInfo.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**